### DISCUSSION TITLE
Janet indexing/method call resolution order potentially confusing

### DISCUSSION THREAD

**Discussion by MikeBeller** [janet-lang/janet#624] [General]
Indexing/method lookup is perhaps a bit confusing in Janet because of the interplay of method resolution and data structure indexing.  To the best of my understanding, the expression `(x y)` is evaluated as:

* If x is a keyword: `((get y x) y)`
* Else, if x is a function or cfunction, this is a regular call `(x y)`
* Else, if x is a collection or abstract: `(get x y)`
* Else: `(get y x)`

I suggest this could be confusing because, for example:

```
(:foo @{:foo (fn [s] 77)})   #=> 77
(:foo @{:foo 77}) #=> nil, but one might have expected 77
```

If someone didn't know about the above resolution order, they might wonder what the heck is going on.

(1) Is this resolution order correct, and is it what was intended and is desired?
(2) If so, should it be documented in one place perhaps, and in what chapter?  (As it crosses between method resolution and data structure indexing.)

-Mike

**Comment by pepe** [comment]
Thanks for enumerating it, I knew about all the cases, but I was not aware of the order.

The example, keyword first/last difference, was discussed many times before and was stated a couple of times it is WNF for @bakpakin. But with rereading it, I guess the second invocation should lead to err?

I very much agree with documenting it somewhere.

**Reply by MikeBeller** [reply]
Thanks @pepe. Not 100% sure whether I got the rules right -- but testing and reading vm.c it seems so.  What does WNF mean?

**Reply by pepe** [reply]
"Will not fix," I believe.

**Reply by pepe** [reply]
And to clarify, I am so used to this mental model for calling method vs. getting from the collection; I can fully understand why it is an important part of Janet.
